coap-gateway/service/client.go

```diff
@@ -883,9 +884,30 @@ func (client *Client) UpdateDeviceMetadata(ctx context.Context, event *events.De
 		client.setShadowSynchronization(sendConfirmCtx, event.GetDeviceId(), previous)
 	}
 	return err
-
 }
 
 func (c *Client) ValidateToken(ctx context.Context, token string) (jwtClaims, error) {
 	return c.server.ValidateToken(ctx, token)
 }
+
+func (c *Client) subscribeToDeviceEvents(owner string, onEvent func(e *authEvents.Event)) error {
+	close, err := c.server.ownerCache.Subscribe(owner, onEvent)
+	if err != nil {
+		return err
+	}
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.closeEventSubscriptions = close
+	return nil
+}
+
+func (c *Client) unsubscribeFromDeviceEvents() {
+	close := func() {}
+	c.mutex.Lock()
+	if c.closeEventSubscriptions != nil {
+		close = c.closeEventSubscriptions
+		c.closeEventSubscriptions = nil
+	}
+	c.mutex.Unlock()
+	close()
```

Comment:
what if `close == nil` ?

Reply:
It's never nil, default value is empty function (`func() {}`) and `c.closeEventSubscriptions` is nil-checked on line 907